### PR TITLE
Masking by bit values

### PR DESCRIPTION
**Pull request by vpipkt**
- [x] function reference docs for new functions
- [x] unit test for SQL snip in docs
- [x] expose extract and mask functions in python api
- [x] ensure python and scala api's are equivalent 
- [x] consider extract bits input and output cell types, unit tests
- [x] update release notes
- [x] Fix https://github.com/locationtech/rasterframes/issues/416 
- [ ] Reorganize masking stuff in RasterFucntionsSpec 
- [x] update the masking page to show landsat mask using mask_by_bits

### CODE REVIEW

**Comment by courtney-whalen** [comment]
I read through the masking doc. I think it looks great!

File: core/src/main/scala/org/locationtech/rasterframes/expressions/package.scala [hunk 1]
```diff
@@ -138,5 +138,7 @@ package object expressions {
     registry.registerExpression[XZ2Indexer]("rf_spatial_index")
 
     registry.registerExpression[transformers.ReprojectGeometry]("st_reproject")
+
+    registry.registerExpression[ExtractBits]("rf_local_extract_bits")
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/package.scala]: Is `rf_mask_by_bit` needed here too?

File: core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/ExtractBits.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  def apply(tile: Column, startBit: Column, numBits: Column): Column =
+    new Column(ExtractBits(tile.expr, startBit.expr, numBits.expr))
+
+  def apply(tile: Tile, startBit: Int, numBits: Int): Tile = {
+    assert(!tile.cellType.isFloatingPoint, "ExtractBits operation requires integral CellType")
+    // this is the last `numBits` positions of "111111111111111"
+    val widthMask = Int.MaxValue >> (63 - numBits)
+    // map preserving the nodata structure
+    tile.mapIfSet(x ⇒ x >> startBit & widthMask)
+  }
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/ExtractBits.scala]: Nicely done.

File: core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala [hunk 3]
```diff
@@ -38,7 +38,15 @@ import org.locationtech.rasterframes.expressions.localops.IsIn
 import org.locationtech.rasterframes.expressions.row
 import org.slf4j.LoggerFactory
 
-abstract class Mask(val left: Expression, val middle: Expression, val right: Expression, inverse: Boolean)
+/** Convert cells in the `left` to NoData based on another tile's contents
+  *
+  * @param left a tile of data values, with valid nodata cell type
+  * @param middle a tile indicating locations to set to nodata
+  * @param right optional, cell values in the `middle` tile indicating locations to set NoData
+  * @param undefined if true, consider NoData in the `middle` as the locations to mask; else use `right` valued cells
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/Mask.scala]: Aha! The insight comes through! Well played, sir!

File: docs/src/main/paradox/reference.md [hunk 4]
```diff
@@ -236,10 +233,35 @@ Generate a `tile` with the values from `data_tile`, with NoData in cells where t
...
+This is a single-bit version of @ref:[`rf_mask_by_bits`](reference.md#rf-mask-by-bits).
+    
+### rf_mask_by_bits
+
+    Tile rf_mask_by_bits(Tile tile, Tile mask_tile, Int start_bit, Int num_bits, Array mask_values) 
+    Tile rf_mask_by_bits(Tile tile, Tile mask_tile, Int start_bit, Int num_bits, list mask_values) 
+    
+Applies a mask from blacklisted bit values in the `mask_tile`. Working from the right, the bits from `start_bit` to `start_bit + num_bits` are @ref:[extracted](reference.md#rf_local_extract_bits) from cell values of the `mask_tile`. In all locations where these are in the `mask_values`, the returned tile is set to NoData; otherwise the original `tile` cell value is returned.
+
+This function is not available in the SQL API. The below is equivalent:
```
Comment (metasim) [inline docs/src/main/paradox/reference.md]: You'll have to explain to me why... If it's array creation, it should be available with the `array(x, y, z, ...)` function.

File: docs/src/main/paradox/reference.md [hunk 5]
```diff
@@ -406,6 +428,15 @@ Returns a `tile` column containing the element-wise inequality of `tile1` and `r
...
 Returns a `tile` column with cell values of 1 where the `tile` cell value is in the provided array or list. The `array` is a Spark SQL [Array][Array]. A python `list` of numeric values can also be passed.
 
+### rf_local_extract_bits
+
+    Tile rf_local_extract_bits(Tile tile, Int start_bit, Int num_bits)
+    Tile rf_local_extract_bits(Tile tile, Int start_bit)
+
+Extract value from specified bits of the cells' underlying binary data. Working from the right, the bits from `start_bit` to `start_bit + num_bits` are extracted from cell values of the `tile`. The `start_bit` is zero indexed. If `num_bits` is not provided, a single bit is extracted.
+
+A common use case for this function is covered by @ref:[`rf_mask_by_bits`](reference.md#rf-mask-by-bits).
```
Comment (metasim) [inline docs/src/main/paradox/reference.md]: I'm really pleased that you broke this whole process down to a set of more fundamental operations. Exactly the approach we want to take, and it's not always obvious to me what those are.

File: pyrasterframes/src/main/python/docs/masking.pymd [hunk 6]
```diff
@@ -14,8 +14,10 @@ spark = create_rf_spark_session()
...
+
+In this section we will demonstrate two common schemes for masking. In Sentinel 2, there is a separate classification raster that defines low quality areas. In Landsat 8, several quality factors are measured and the indications are packed into a single integer, which we have to unpack.
  
-## Masking Example
+## Masking Sentinel 2
 
 Let's demonstrate masking with a pair of bands of Sentinel-2 data. The measurement bands we will use, blue and green, have no defined NoData. They share quality information from a separate file called the scene classification (SCL), which delineates areas of missing data and probable clouds. For more information on this, see the [Sentinel-2 algorithm overview](https://earth.esa.int/web/sentinel/technical-guides/sentinel-2-msi/level-2a/algorithm). Figure 3 tells us how to interpret the scene classification. For this example, we will exclude NoData, defective pixels, probable clouds, and cirrus clouds: values 0, 1, 8, 9, and 10.
```
Comment (metasim) [inline pyrasterframes/src/main/python/docs/masking.pymd]: Add a `@@@Note` block here indicating GDAL is required for this example. See the `getting-started.pymd` for the syntax if needed.

File: pyrasterframes/src/main/python/docs/masking.pymd [hunk 7]
```diff
@@ -98,9 +100,72 @@ masked.withColumn('green_masked', rf_mask(rf_convert_cell_type('green', masked_b
...
+                  rf_mask_by_bits('data_masked', 'mask', 11, 2, [2, 3])) \
+       .withColumn('data2', # mask other data col against the other band
+                  rf_mask(rf_convert_cell_type('data2',CellType.uint16()), 'data_masked')) \
+       .filter(rf_data_cells('data_masked') > 0) # remove any entirely ND rows
+
+# Inspect a sample
+l8_df.select('data', 'mask', 'data_masked', 'data2', rf_extent('data_masked')) \
+     .filter(rf_data_cells('data_masked') > 32000)
+```
+
```
Comment (metasim) [inline pyrasterframes/src/main/python/docs/masking.pymd]: Wow. That's all I can say. This is absolutely brilliant.

File: pyrasterframes/src/main/python/tests/PyRasterFramesTests.py [hunk 8]
```diff
@@ -139,6 +141,22 @@ def test_tile_udt_serialization(self):
             long_trip = df.first()["tile"]
             self.assertEqual(long_trip, a_tile)
 
+    def test_masked_deser(self):
```
Comment (metasim) [inline pyrasterframes/src/main/python/tests/PyRasterFramesTests.py]: Not sure I understand what this is testing...

File: pyrasterframes/src/main/python/tests/RasterFunctionsTests.py [hunk 9]
```diff
@@ -256,28 +256,77 @@ def test_mask_by_values(self):
...
+
+        self.assertTrue(
+            mask_fill_df.select(rf_data_cells('mbb')).first()[0],
+            16 - 4
+        )
+        # Unsure why this fails. mask_fill_tile.cells is all 42 unmasked.
+        # self.assertEqual(mask_fill_tile.cells.mask.sum(), 4,
+        #                  f'Expected {16 - 4} data values but got the masked tile:'
+        #                  f'{mask_fill_tile}'
+        #                  )
```
Comment (metasim) [inline pyrasterframes/src/main/python/tests/RasterFunctionsTests.py]: Are you expecting `mask_fill_tile.cells.mask.sum()` to return `42*4*4`, or the `cells.sum()` after the mask has been applied?

Comment (metasim) [reply pyrasterframes/src/main/python/tests/RasterFunctionsTests.py]: Or something else?

File: core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/ExtractBits.scala (hunk 2 above)
Comment (vpipkt) [reply core/src/main/scala/org/locationtech/rasterframes/expressions/transformers/ExtractBits.scala]: 🙇

File: docs/src/main/paradox/reference.md (hunk 4 above)
Comment (vpipkt) [reply docs/src/main/paradox/reference.md]: Basically I threw in the towel on trying to create an Expression with 5 arguments (tile, tile, int, int, array) ... If you have any intiution on this it's welcome! Maybe we can file another issue to bring that along.

File: pyrasterframes/src/main/python/docs/masking.pymd (hunk 6 above)
Comment (vpipkt) [reply pyrasterframes/src/main/python/docs/masking.pymd]: So the files are actually GeoTiffs (COGs I think) on our test bucket.

File: pyrasterframes/src/main/python/tests/PyRasterFramesTests.py (hunk 8 above)
Comment (vpipkt) [reply pyrasterframes/src/main/python/tests/PyRasterFramesTests.py]: basically just testing that a Tile that gets deserialized has the expected masking on it.

File: pyrasterframes/src/main/python/tests/RasterFunctionsTests.py (hunk 9 above)
Comment (vpipkt) [reply pyrasterframes/src/main/python/tests/RasterFunctionsTests.py]: so the mask is binary where 1 indicates that location is nodata.

So by doing `.mask.sum()` we expect the same as `rf_no_data_cells`. But i kept getting this assertion failing locally.

Comment (vpipkt) [reply pyrasterframes/src/main/python/tests/RasterFunctionsTests.py]: so the mask is binary where 1 indicates that location is nodata.

So by doing .mask.sum() we expect the same as rf_no_data_cells. But i kept getting this assertion failing locally.

File: pyrasterframes/src/main/python/tests/PyRasterFramesTests.py (hunk 8 above)
Comment (vpipkt) [reply pyrasterframes/src/main/python/tests/PyRasterFramesTests.py]: Basically is the same thing as my commented out block at RasterFunctionsTest.py lines 282 on.... paranoia

File: pyrasterframes/src/main/python/docs/masking.pymd (hunk 6 above)
Comment (metasim) [reply pyrasterframes/src/main/python/docs/masking.pymd]: Aha.

**Pull request merged**